pkg/astrolabe/pe_zip.go

```diff
@@ -20,26 +20,196 @@ import (
+	panic("implement me")
+}
+
+func (recv ZipProtectedEntity) ListSnapshots(ctx context.Context) ([]ProtectedEntitySnapshotID, error) {
+	panic("implement me")
+}
+
+func (recv ZipProtectedEntity) DeleteSnapshot(ctx context.Context, snapshotToDelete ProtectedEntitySnapshotID, params map[string]map[string]interface{}) (bool, error) {
+	panic("implement me")
+}
+
+func (recv ZipProtectedEntity) GetInfoForSnapshot(ctx context.Context, snapshotID ProtectedEntitySnapshotID) (*ProtectedEntityInfo, error) {
+	panic("implement me")
+}
+
+func (recv ZipProtectedEntity) GetComponents(ctx context.Context) ([]ProtectedEntity, error) {
+	return recv.components, nil
+}
+
+func (recv ZipProtectedEntity) GetID() ProtectedEntityID {
+	return recv.info.GetID()
+}
+
+func (recv ZipProtectedEntity) GetDataReader(ctx context.Context) (io.ReadCloser, error) {
+	return recv.data.Open()
+}
+
+func (recv ZipProtectedEntity) GetMetadataReader(ctx context.Context) (io.ReadCloser, error) {
+	return recv.metadata.Open()
+}
+
+func (recv ZipProtectedEntity) Overwrite(ctx context.Context, sourcePE ProtectedEntity, params map[string]map[string]interface{}, overwriteComponents bool) error {
+	panic("implement me")
+}
+
+func GetPEFromZipStream(ctx context.Context, reader io.ReaderAt, size int64) (ProtectedEntity, error) {
+	zipReader, err := zip.NewReader(reader, size)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not init zip file")
+	}
+
+	retEntity := ZipProtectedEntity{
+		combined:   reader,
+		components: []ProtectedEntity{},
+	}
+
+	for _, checkFile := range zipReader.File {
+		if strings.HasSuffix(checkFile.Name, ".peinfo") {
+			peinfoReader, err := checkFile.Open()
+			if err != nil {
+				return nil, errors.Wrapf(err, "could not open %s", checkFile.Name)
+			}
+			peinfoBuf, err := ioutil.ReadAll(peinfoReader)
+			if err != nil {
+				return nil, errors.Wrapf(err, "could not read data for %s", checkFile.Name)
+			}
+			peinfo := ProtectedEntityInfoImpl{}
+			err = json.Unmarshal(peinfoBuf, &peinfo)
+			if err != nil {
+				return nil, errors.Wrapf(err, "could not unmarshal JSON from %s", checkFile.Name)
+			}
+			retEntity.info = peinfo
+		}
+		if strings.HasSuffix(checkFile.Name, ".md") {
+			retEntity.metadata = checkFile
+			continue
+		}
+		if strings.HasSuffix(checkFile.Name, "data") {
+			retEntity.data = checkFile
+			continue
+		}
+		if strings.HasPrefix(checkFile.Name, "/components") {
+			dataOffset, err := checkFile.DataOffset()
+			if err != nil {
+				return nil, errors.Wrapf(err, "Could not get data offset for %s", checkFile.Name)
+			}
+			componentReader := io.NewSectionReader(reader, dataOffset, int64(checkFile.CompressedSize64))
+			componentPE, err := GetPEFromZipStream(ctx, componentReader, int64(checkFile.CompressedSize64))
+			retEntity.components = append(retEntity.components, componentPE)
+			continue
+		}
+		// TODO - Should log an error here for unknown type
```

Comment:
Are you going to add a log here in this PR?

Reply:
Not yet, I'd have to have a logger come in from somewhere, so I'm going to see if we use this thing more.